src/theory/quantifiers/sygus/sygus_reconstruct.h

```diff
@@ -0,0 +1,358 @@
+
+  /**
+   * Remove solved obligations from `d_unsolvedObs.
+   */
+  void removeSolvedObs();
+
+  /**
+   * Replace all bound variables in `n` with ground values of the type.
+   *
+   * @param n A term containing bound variables
+   * @return `n` with all vars in `n` replaced with ground values
+   */
+  Node replaceVarsWithGroundValues(Node n) const;
+
+  /**
+   * A notification that s is equal to n * { vars -> subs }. This function
+   * should return false if we do not wish to be notified of further matches.
+   */
+  bool notify(Node s,
+              Node n,
+              std::vector<Node>& vars,
+              std::vector<Node>& subs) override;
+
+  /**
+   * Reset the state of this SygusReconstruct object
+   */
+  void reset();
+
+  /**
+   * Return a string representation of an obligation.
+   *
+   * @param k An obligation
+   * @return A string representation of `k`
+   */
+  std::string ob(Node k) const;
+
+  /**
+   * Print all reachable obligations and their candidate solutions from
+   * the `root` obligation and its candidate solutions.
+   *
+   * \note requires enabling "sygus-rcons" trace
+   *
+   * @param root The root obligation to start from
+   */
+  void printCandSols(const Node& mainOb) const;
+
+  /**
+   * Print the pool of patterns/shape used in the matching phase.
+   *
+   * \note requires enabling "sygus-rcons" trace
+   *
+   * @param pool a pool of patterns/shapes to print
+   */
+  void printPool(const std::unordered_map<TypeNode,
+                                          std::vector<Node>,
+                                          TypeNodeHashFunction>& pool) const;
+
+  /** pointer to the sygus term database */
+  TermDbSygus* d_tds;
+  /** reference to the statistics of parent */
+  SygusStatistics& d_stats;
+
+  /** a set of obligations that are not yet satisfied for each sygus datatype */
+  std::unordered_map<TypeNode,
+                     std::unordered_set<Node, NodeHashFunction>,
+                     TypeNodeHashFunction>
+      d_unsolvedObs;
+  /** a map from an obligation to the intended builtin term */
+  std::unordered_map<Node, Node, NodeHashFunction> d_builtinTerm;
+  /** a map from a builtin term to its obligation (inverse of above map). Unlike
+   * in the map above, the sygus type of the obligation is needed */
+  std::unordered_map<TypeNode,
+                     std::unordered_map<Node, Node, NodeHashFunction>,
+                     TypeNodeHashFunction>
+      d_ob;
+  /** a map from an obligation to its sygus solution (if it exists) */
+  std::unordered_map<TNode, TNode, TNodeHashFunction> d_sol;
+  /** a map from an obligation to a set of its candidate sygus solutions */
+  std::unordered_map<Node,
+                     std::unordered_set<Node, NodeHashFunction>,
+                     NodeHashFunction>
+      d_candSols;
+
+  /** a map from a candidate solution to its sub-obligations */
+  std::unordered_map<Node, std::vector<Node>, NodeHashFunction> d_subObs;
```

Comment:
As I mentioned, I think it would be helpful to group the ranges of the maps with the same domain to an info class.  Here is an example of a refactoring PR which consolidated many parallel maps into a SygusTypeInfo class: https://github.com/CVC4/CVC4/pull/3187/files.

Reply:
I have added an `ObligationInfo` class. I thought about adding another class for maps with sygus type as the domain, but I didn't see any other relationship between them justify it.

Reply:
Ok sounds good.